src/plugins/kibana_utils/public/state_containers/create_state_container.ts

```diff
@@ -20,34 +20,52 @@
 import { BehaviorSubject } from 'rxjs';
 import { skip } from 'rxjs/operators';
 import { RecursiveReadonly } from '@kbn/utility-types';
+import deepFreeze from 'deep-freeze-strict';
 import {
   PureTransitionsToTransitions,
   PureTransition,
   ReduxLikeStateContainer,
   PureSelectorsToSelectors,
+  BaseState,
 } from './types';
 
 const $$observable = (typeof Symbol === 'function' && (Symbol as any).observable) || '@@observable';
+const $$setActionType = '@@SET';
 
 const freeze: <T>(value: T) => RecursiveReadonly<T> =
   process.env.NODE_ENV !== 'production'
     ? <T>(value: T): RecursiveReadonly<T> => {
-        if (!value) return value as RecursiveReadonly<T>;
-        if (value instanceof Array) return value as RecursiveReadonly<T>;
-        if (typeof value === 'object') return Object.freeze({ ...value }) as RecursiveReadonly<T>;
-        else return value as RecursiveReadonly<T>;
+        const isFreezable = value !== null && typeof value === 'object';
+        if (isFreezable) return deepFreeze(value) as RecursiveReadonly<T>;
+        return value as RecursiveReadonly<T>;
       }
     : <T>(value: T) => value as RecursiveReadonly<T>;
 
-export const createStateContainer = <
-  State,
+export function createStateContainer<State extends BaseState>(
+  defaultState: State
+): ReduxLikeStateContainer<State>;
+export function createStateContainer<State extends BaseState, PureTransitions extends object>(
+  defaultState: State,
+  pureTransitions: PureTransitions
+): ReduxLikeStateContainer<State, PureTransitions>;
+export function createStateContainer<
```

Comment:
I am not exactly sure, but I ended up adding overrides here because bumped into following with previous setup just with optional generics:

```
const container = createStateContainer({test: 'test}, {action: (s) => () => s}); 
container.get().test // ts works
container.transitions.action() // ts works 
```

but
explicitly specify only State shape in generic:
```
type S = {test: string};
const container = createStateContainer<S>({test: 'test}, {action: (s) => () => s}); 
container.get().test // ts works
container.transitions.action() // doesn't work.. ts now think, that transitions are {}. 
```

So adding those overrides instead of using optional generic params just made it explicit, that when generic params are specified explicitly - the number of specified params have to match number of arguments.